### ISSUE TITLE
String formatting in actually portable Awk, printf "%1s"

### ISSUE THREAD

**Issue by lukq** [jart/cosmopolitan#879]
I hope this is the correct place to report an issue with actually portable Awk. I use the version of the program awk-2022-08-22.com downloaded from https://justine.lol/awk/ . I do not have a Linux environment, so I just use the compiled binary. I run it under Windows 10. I found a behaviour of string formatting with printf and %s that may be a bug.

I have the following code saved in file stmt:
`BEGIN{printf"(%1s)","\r"}`

I run this command from command line:
`awk-2022-08-22.com -f stmt >b`

I expect the contents of b to be three bytes. The actual result I get is 4 bytes.
The bytes in hex code are 28 20 0D 29.

It seems as if the string "\r" is seen to have length 0, so the padding of one space is added to have the requested width for %1s. I think that is a bug. I observed it on Windows, I did not test on other oparating systems.

My use case is generating an output file with fixed length fields, where strings in fields should be terminated by the \r character, so I would use it like printf "%16s","text\r".

**Comment by jart** [comment]
It's not strictly standard C, but this might be a feature rather than a bug.

Cosmopolitan's `printf()` uses `wcwidth()` under the hood so that directives like `%16s` take UNICODE CJK monospace width into consideration, so you don't end up with broken tables due to having combining marks or emoji, etc.

You say your use case is `printf "%16s","text\r"`. Your `\r` is being properly included at the end. Looking at that code, I actually wouldn't want the `\r` to take up a character slot in the `%16s` because the purpose of `\r` is to rewind the cursor back to the beginning of the line. Are you sure you want it to be packed into your fields?

**Comment by lukq** [comment]
Ok, I see. If that is a feature of this implementation, then I have to write the output in a different way. I was not aware of this, that's why I assumed it might be a bug.

I can give some more details about the problem I was solving. It is probably not a typical use of Awk. For some testing I needed to generate a binary file with the contents of a casette tape for an 8 bit computer emulator. The tape file contains 128-byte header followed by variable length data. The header has several fields that start at fixed offsets and have fixed byte lengths. So, first byte is file type, then 16 bytes are file name, and then the following fields give the load address, file length, execution address and comment. The file name is exactly 16  bytes, so if the file name is exactly 16 characters, you write those characters, if it is 15 or less characters, you write the characters, then \r to indicate the end of the name, and then any padding characters to fill up to 16 bytes. I thought I have a clever solution by appending \r to the name and writing all with %16s format, but that does not work. If the name is 16 characters, I need the \r to not be printed, and if the name is less that 15 characters, I need the \r to be printed but also padding spaces to be added to fill to exactly 16 bytes.

I know Awk is typically not used to write binary files. Some Awk implementations handle only text files well, some handle also binary files without problem. Important is the ability to write characters with values 0, 10, and 13, those are the problematic ones. The cosmopolitan Awk works fine with binary output, except for this %16s formatting.

I have solved my problem in some other way, so I don't need the formatting behavior to be changed, now that I know how it works. Thanks for your reply.
